Enhance the OpenAPI plugin to accept user-provided values for parameters

artem-smotrakov opened the pull request:
This is a patch for #17315 

Main changes:

- Added `parameter_values_location` configuration option for the OpenAPI plugin. The option allows to set a path to a YAML file which contains context-specific parameter values.
- Added `ParameterValues` class which can load context-specific parameter values from a YAML file. The file contains a mapping `{ path, parameter name } -> list of values` . Currently, it doesn't use an HTTP method but it can be easily updated. For `in: body` parameter type, a user can specify a JSON payload, but it's not currently possible to specify a value for a particular field in the JSON payload.
- Updated `ParameterHandler` to use context-specific parameter values. If a user provided custom parameter values, then the class prefers them, and tries to enumerate all possible combinations of the specified parameters.
- Fixed a possible null-dereference in `get_uri()` method.
- Added a couple of tests.

I am testing the feature on a couple of applications but would like to start a review now to get feedback earlier.

artem-smotrakov committed:
Added parameter_values_location configuration parameter for open_api plugin

artem-smotrakov committed:
Try to avoid NPE in RequestFactory.get_uri() method

artem-smotrakov committed:
Update the OpenAPI plugin to accept user-provided values for parameters

artem-smotrakov committed:
Added a couple of tests for setting custom parameter values for API endpoints

artem-smotrakov committed:
Describe new methods for setting context-specific parameter values

artem-smotrakov committed:
Use correct parameter names in tests for the OpenAPI plugin

artem-smotrakov committed:
open_api should carefully check if self._parameter_values_location: was set

artem-smotrakov committed:
Merge remote-tracking branch 'upstream/develop' into openapi-parameter-values

artem-smotrakov commented:
I have tested it with a couple of applications - it worked well.

andresriancho commented on w3af/core/data/parsers/doc/open_api/parameters.py:
```diff
@@ -33,6 +35,83 @@ class OpenAPIParamResolutionException(Exception):
+        Loads parameter values from YAML.
+        :param string: Definition of parameter values in YAML.
+        """
+        content = yaml.load(string)
```
User might enter an invalid YAML, that case should be handled.

artem-smotrakov replied:
Okay, I'll add a try-except block to catch `YAMLError` and wrap the exception in to `ParameterValueParsingError`.

andresriancho commented on w3af/plugins/crawl/open_api.py:
```diff
@@ -407,6 +415,13 @@ def get_options(self):
+        h = ('This option sets a path to a YAML file which contains parameter values'
+             ' which should be used in testing API endpoints. If no parameter values are provided,'
+             ' the plugin tries to guess them.')
+        o = opt_factory('parameter_values_location', self._parameter_values_location, d, INPUT_FILE, help=h)
```
Maybe the YAML format validation should be done here? That could be done by changing the `INPUT_FILE` type to `YAML_INPUT_FILE` and creating the logic for handling that.

It should be fairly easy, take a look at:
 * https://github.com/andresriancho/w3af/blob/master/w3af/core/data/options/input_file_option.py
 * https://github.com/andresriancho/w3af/blob/356b14b975039706f4fd7f4f5db5b114cd75f14e/w3af/core/data/options/tests/test_input_file_option.py
 * https://github.com/andresriancho/w3af/blob/356b14b975039706f4fd7f4f5db5b114cd75f14e/w3af/core/data/options/opt_factory.py#L43

artem-smotrakov replied:
Hmm, that may be nice. Although `input_file_option.py` doesn't look that simple at first glance :) Okay, I'll see what I can do here.

andresriancho replied:
Yeah, `input_file_option.py` doesn't look simple, but `yaml_file_option.py` should be simple :-)

Take a look at the other files in the same directory, those will give you ideas on how simple the inputs can actually be.

andresriancho commented on w3af/plugins/crawl/open_api.py:
```diff
@@ -407,6 +415,13 @@ def get_options(self):
+        h = ('This option sets a path to a YAML file which contains parameter values'
+             ' which should be used in testing API endpoints. If no parameter values are provided,'
+             ' the plugin tries to guess them.')
+        o = opt_factory('parameter_values_location', self._parameter_values_location, d, INPUT_FILE, help=h)
```
`parameter_values_location` should be changed to `parameter_values_file` (that is how input files are usually named)

https://github.com/andresriancho/w3af/blob/356b14b975039706f4fd7f4f5db5b114cd75f14e/w3af/plugins/output/text_file.py#L297
https://github.com/andresriancho/w3af/blob/356b14b975039706f4fd7f4f5db5b114cd75f14e/w3af/plugins/output/xml_file.py#L153

andresriancho replied:
Yes please

andresriancho commented on w3af/plugins/crawl/open_api.py:
```diff
@@ -448,10 +464,35 @@ def get_long_desc(self):
+                - 1234567
+              - name: birth-date
+                values:
+                - 2000-01-02
```
I believe that this feature requires its own documentation in https://github.com/andresriancho/w3af/blob/develop/doc/sphinx/scan-rest-apis.rst

I would make the example in `get_long_desc` shorter, and reference users to the documentation with something like: "For more information and examples about this feature read the framework documentation at https://docs.w3af.org"

artem-smotrakov replied:
Okay, I'll update both description and docs. In fact, the docs need to be updated with descriptions of the recent changes/features. I hope I can find time to do that.

andresriancho reviewed (commented):
Awesome!

andresriancho commented:
I need to work harder on merging your PR. If I don't do that I'll end up in a branch merge nightmare.

Sorry for not being more responsive with these PRs. I was on vacations. Will try to merge all these amazing things you've been working on to `develop` as soon as the comments are solved.

artem-smotrakov commented:
Thanks for your comments. Let's work on them one by one, and I'll resolve possible conflicts which may appear. Then, it should be easier to merge them to `develop`.

artem-smotrakov committed:
Rename parameter_values_location to parameter_values_file

artem-smotrakov committed:
Expect invalid yaml in ParameterValues

artem-smotrakov committed:
Merge

artem-smotrakov committed:
Better documentation and description for 'parameter_values_file' parameter of the crawl.open_api plugin

artem-smotrakov committed:
Rename 'custom_spec_location' parameter to 'custom_spec_file' in the crawl.open_api plugin

artem-smotrakov committed:
Added YAML_INPUT_FILE option type

artem-smotrakov committed:
Check for an empty value in YamlFileOption class

artem-smotrakov committed:
Use YAML_INPUT_FILE for 'parameter_values_file' option in the crawl.open_api plugin

artem-smotrakov committed:
Added TestOpenAPICustomSpec

artem-smotrakov committed:
Added TestOpenAPICustomParameterValues

artem-smotrakov committed:
Use common by_path() in test_open_api.py

artem-smotrakov commented:
I have updated the patch:

- Renamed configuration parameters in the `crawl.open_api` plugin.
- Added `YAML_INPUT_FILE` option type.
- Updated docs and description for the `crawl.open_api` plugin.
- The `crawl.open_api` plugin now expects an invalid YAML file.
- Added `TestOpenAPICustomParameterValues` for the new configuration option.
- Added `TestOpenAPICustomSpec` for `custom_spec_file` configuration option.

artem-smotrakov commented:
@andresriancho When you have  some time, could you please take a look at this pull request? I have also updated other pull requests and added several comments - looking for a review. Thanks!
